driver-core/src/main/com/mongodb/AutoEncryptionSettings.java

```diff
@@ -265,12 +281,37 @@ public String getKeyVaultNamespace() {
      *     <li>key: byte[] of length 96, the local key</li>
      * </ul>
      *
+     * <p>
+     * It is also permitted for the value of a kms provider to be an empty map, in which case the driver will first
+     * </p>
+     * <ul>
+     *  <li>use the {@link Supplier} configured in {@link #getKmsProviderSupplierMap()} to obtain a non-empty map</li>
+     *  <li>attempt to obtain credentials from the environment</li>
+     * </ul>
+     *
      * @return map of KMS provider properties
+     * @see #getKmsProviderSupplierMap()
      */
     public Map<String, Map<String, Object>> getKmsProviders() {
         return unmodifiableMap(kmsProviders);
     }
 
+    /**
+     * Gets the KMS provider to Supplier map.
+     *
+     * <p>
+     * If the {@link #getKmsProviders()} map contains an empty map as its value, the driver will use a {@link Supplier} configured for
+     * the same provider in this map to obtain a non-empty map that contains the credential for the provider.
```

Comment:
The [current behavior](https://github.com/mongodb/libmongocrypt/blob/c751982a12fff7a54d83909da3a4ede450970f39/src/mongocrypt.h.in#L482-L484) of libmongocrypt is to only enter the `MONGOCRYPT_CTX_NEED_KMS_CREDENTIALS` if an empty `aws: {}` document was set in  `mongocrypt_setopt_kms_providers`. This suggests that users can provide credentials for other KMS providers.

I think it is a small change in libmongocrypt to enter the `MONGOCRYPT_CTX_NEED_KMS_CREDENTIALS` state if other KMS providers are configured with an empty document. I prefer that over to documenting the limitation with "aws", and can try to make that change today.

The open question I have is: Should users be permitted to provide credentials for "local", or "kmip"? There may not be a use case. But it is more consistent with this proposed API.

Reply:
It seems like the cleanest way to allow it for all.